File: CSharp/[Udemy] CSharp POO/session13/Generics_Set_Dictionary/Generics_2_Restrictions/Entities/Product.cs
using System.Globalization;

namespace Generics_2_Restrictions.Entities
{
    internal class Product : IComparable
    {
        public string Name { get; set; }
        public double Price { get; set; }
        public List<Product> list = new List<Product>();

        public Product() { }
        public Product(string name, double price)
        {
            Name = name;
            Price = price;            
        }
        public void AddProduct(Product prod)
        {
            list.Add(prod);
        }
        public void RemoveProduct(Product prod)
        {
            list.Remove(prod);
        }
        public override string ToString()
        {
            return $"{Name}, {Price.ToString("F2", CultureInfo.InvariantCulture)}";
        }

        public int CompareTo(object? obj)
        {
            if (obj is not Product)
            {
                throw new ArgumentException("Comparing error: object is not a Product.");
            }
            
            Product other = obj as Product;

            return Price.CompareTo(other.Price);
        }
    }
}

File: Java/[Udemy] Java POO/section8/aula67/Product.java
package aula67;

public class Product {
	
	public String Name;
	public double Price;
	public int Quantity;
	
	public Product() {}
	
	public double TotalValueInStock() {
		return Price * Quantity;
	}
	
	public void AddProduct(int quantity) {
		Quantity += quantity;
	}
	
	public void RemoveProduct(int quantity) {
		Quantity -= quantity;
	}
	
	public String toString() {
		return String.format("%s, $ %.2f, %d units, Total: $ %.2f%n", Name, Price, Quantity, TotalValueInStock());
	}

}